iscsi: don't write json medata file when the volume is already mounted.

jsafrane opened the pull request:
iSCSI volume plugin persists volume metadata into global mount directory, before it is mounted. Content of the directory is shadowed by the volume mount. Therefore kubelet should not write metadata to the directory when a second pod uses the same volume on the same node.
    
1. The metadata were already persisted before mounting the volume for the first pod.
2. The global mount directory has the volume mounted, so any write there would write to the volume, which is undesirable.

Fixes: #91736

```release-note
Fixed using of a read-only iSCSI volume in multiple pods.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jsafrane committed:
Add e2e test for read-only volume used by multiple pods

jsafrane commented:
/hold cancel

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
Can't find what's wrong with `Pre-provisioned PV (default fs)] subPath should support existing directories when readOnly specified in the volumeSource ` - the same test passes for in-line volume. And on my cluster.

jsafrane commented:
I think I know what's going on: kernel iSCSI target looses data.  In kubelet logs I can see that the faulty test creates a pod with subPath, that gets correctly created, but them it run she same pod with the same subPath as read-only. Kubelet should not create the subpath dir (it already exists from the old pod), however, this subdir is not present on the volume.

Similarly, "volumes should store data" tests fails for iscsi from time to time, when it misses one write() from a pod that created test data on the volume few seconds before.

https://bugzilla.redhat.com/show_bug.cgi?id=1721218

We should rework the tests for some more stable iSCSI target implementation.

jsafrane commented:
/retest

bertinatto commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -623,13 +654,22 @@ func (util *ISCSIUtil) DetachDisk(c iscsiDiskUnmounter, mntPath string) error {
 	}
+	err = util.removeISCSI(mntPath)
+	if err != nil {
+		return fmt.Errorf("failed to remove metadata file: %s", err)
```
nit: `%s` while (it seems like) existing code uses `%v`

jsafrane replied:
It's really mixed in the code, but I added `%v` to couple of places.

bertinatto commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -204,34 +204,63 @@ func (util *ISCSIUtil) MakeGlobalVDPDName(iscsi iscsiDisk) string {
+		if !os.IsNotExist(openerr) {
+			return fmt.Errorf("iscsi: failed to open %s: %s", jsonFilePath, openerr)
+		}
+		// JSON file does not exist. For backward compatibility: try iscsi.json in the mount directory.
```
For how long do we want to be backward-compatible with `iscsi.json`? Can we add a TODO to remove this v1.2x?

jsafrane replied:
TODO added

bertinatto commented:
/assign @bswartz

bertinatto commented:
Changes LGTM; will let other people review as well.

CC @j-griffith

jsafrane commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -204,34 +204,63 @@ func (util *ISCSIUtil) MakeGlobalVDPDName(iscsi iscsiDisk) string {
+		fp, err = os.Open(file)
+		if err != nil {
+			// Report original openerror
+			return fmt.Errorf("iscsi: failed to open %s: %s", file, openerr)
```
It's explicitly commented that `openerr` is returned. The error from the "new" file is IMO more interesting than the legacy one.

bswartz commented:
I have a high level question about this fix. Why can't we use a different mechanism to avoid reading/writing the json file in the case when the volume is already attached to the current node?

It seems like if we knew the volume was attached already when attaching to the second pod, we could skip the writing of the json file. And if we knew the volume was still attached to another pod while detaching from a pod, we couid skip the reading of the file. The contents of the file should only be necessary to do the final detach, and in that case the json file should never be shadowed by a mount.

I might be missing something about what we use the file contents for, but my understanding was that we only needed it to perform the logout, and we had a different way to do reference counting so we can reliably tell if we're not the first pod to attach to a given volume, and if we're not the last pod to detach from a given volume.

jsafrane committed:
iscsi: don't write json medata file when the volume is already mounted.

iSCSI volume plugin persists volume metadata into global mount directory,
before it is mounted. Content of the directory is shadowed by the volume
mount. Therefore kubelet should not write metadata to the directory when a
second pod uses the same volume on the same node.

1. The metadata were already persisted before mounting the volume for the
first pod.
2. The global mount directory has the volume mounted, so any write there
would write to the volume, which is undesirable.

jsafrane commented:
> The contents of the file should only be necessary to do the final detach, and in that case the json file should never be shadowed by a mount.

@bswartz, you're correct! I reworked the patch not to move the location of iSCSI metadata file and write it only for the first time. Please check the new PR descripion & code.

bswartz reviewed (approved):
/lgtm

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/test pull-kubernetes-e2e-gce-iscsi-serial

jsafrane commented:
/assign @andrewsykim @oomichi 
for framework change approval

oomichi commented:
e2e framework part also seems good for me.

/approve

jsafrane commented:
/king bugfix

jsafrane commented:
/kind bug

jsafrane commented:
/retest
